sotodlib/io/bookbinder.py

```diff
@@ -571,6 +582,37 @@ def flush(self, flush_time=None):
         if smurf_primary is not None:
             f['primary'] = smurf_primary
 
+        # Ancillary data object to be written to disk (co-sampled HK encoder data)
+        anc_data = core.G3TimesampleMap()
+        anc_data.times = smurf_data.times
+
+        ## Adds HWP data if present
+        if self.hwp_loader is not None:
+            if self.hwp_loader.hwp_angle_interp is not None:
+                ts = np.array(smurf_data.times) / core.G3Units.s
+                hwp_angle = self.hwp_loader.hwp_angle_interp(ts)
+                hwp_angle = np.mod(hwp_angle, 2*np.pi)
+
+                hwp_locked = self.hwp_loader.locked_interp(ts)
+                hwp_stable = self.hwp_loader.stable_interp(ts)
+                hwp_rate = self.hwp_loader.stable_interp(ts)
+
+                flags = np.zeros_like(hwp_angle, dtype=np.int32)
+                m = hwp_rate < MIN_HWP_RATE
+                flags[m] = np.bitwise_or(flags[m], HWPFlags.STATIONARY)
+                m = hwp_locked != 1
+                flags[m] = np.bitwise_or(flags[m], HWPFlags.UNLOCKED)
+                m = hwp_stable != 1
+                flags[m] = np.bitwise_or(flags[m], HWPFlags.UNSTABLE)
+                m = np.isnan(hwp_angle)
+                flags[m] = np.bitwise_or(flags[m], HWPFlags.NO_DATA)
+
+                hwp_angle[flags != 0] = -flags[flags != 0]
+            else:
+                hwp_angle = np.full(len(smurf_data.times), -HWPFlags.NO_DATA)
+
```

Comment:
Perhaps this should be in its own method (e.g., `get_hwp_angles`), to avoid having too much in `flush`.

Reply:
Sounds good, I can add this in a method of the HWPLoader